File: src/inventory.py
import random

from config.settings import SETTINGS
from src.Helpers import storable_items
from src.ui import display_inventory, display_success_message, display_error_message, display_warning_message


class Inventory:
    def __init__(self):
        """Initialize an empty inventory."""
        self.items = {}
        self.generate_random_items()

    def add_item(self, item, quantity=1):
        """Add an item to the inventory."""
        if item in self.items:
            self.items[item] += quantity
        else:
            self.items[item] = quantity
    def show_inventory(self):
        """Display inventory contents."""
        display_inventory(self.items)

    def use_item(self, item, player):
        """Use an item from inventory."""
        if item in self.items.keys() and self.items[item] > 0:
            if item == "polttoaine":
                if player.fuel + SETTINGS['fuel_can_capacity'] > SETTINGS['max_fuel']:
                    player.fuel = SETTINGS['max_fuel']
                else:
                    player.fuel += SETTINGS['fuel_can_capacity']
            elif item == "lääke":
                if player.health + SETTINGS['medicine_health'] > SETTINGS['max_health']:
                    player.health = SETTINGS['max_health']
                else:
                    player.health += SETTINGS['medicine_can_health']
            elif item == "ruoka":
                if player.health + SETTINGS['food_can_health'] > SETTINGS['max_health']:
                    player.health = SETTINGS['max_health']
                else:
                    player.health += SETTINGS['food_can_health']
            elif item == "vesi":
                if player.health + SETTINGS['water_can_health'] > SETTINGS['max_health']:
                    player.health = SETTINGS['max_health']
                else:
                    player.health += SETTINGS['water_can_health']
            elif item == "ase":
                display_warning_message("Aseita ei voi käyttää suoraan, ne käytetään automaattisesti puolustautumaan zombien hyökkäyksiltä.")
                return
            self.items[item] -= 1
            display_success_message(f"Käytetty {item}. Jäljellä: {self.items[item]}")
        else:
            display_error_message(f"{item} ei saatavilla.")
    def generate_random_items(self):
        """Generate random items at the start"""
        random_inventory_items = random.sample(storable_items, 3)
        for item in random_inventory_items:
            self.add_item(item, random.randint(1, 3))

File: frontend/js/game/inventory.js
import {storableItems} from "./helpers.js";
import {SETTINGS} from "./settings.js";
import {showSnackbar, snackbarType} from "../components/snackbar.js";
import {playSoundEffect, soundEffects} from "../components/sound_effects.js";
import {updateInventoryUI, updateUI} from "../components/ui_handler.js";

export default class Inventory {
    constructor(items = {}) {
        /** Initialize an empty inventory. */
        this.items = items;
        if(Object.keys(this.items).length === 0) {
            this.generateRandomItems();
        }
    }

    addItem(item, quantity = 1) {
        /** Add an item to the inventory. */
        if (this.items[item]) {
            this.items[item] += quantity;
        } else {
            this.items[item] = quantity;
        }
    }

    loadInventory(loadedInventory) {
        /** Load a saved inventory. */
        this.items = loadedInventory.items;
    }
    useItem(item, player) {
        /** Use an item from inventory. */
        if (this.items[item] && this.items[item] > 0) {
            switch (item) {
                case "fuel":
                    if (player.fuel < SETTINGS.max_fuel) {
                        playSoundEffect(soundEffects.fuel)
                        player.fuel = Math.min(player.fuel + SETTINGS.fuel_can_capacity, SETTINGS.max_fuel);
                    }else {
                        playSoundEffect(soundEffects.ERROR)
                        showSnackbar(snackbarType.ERROR, "Fuel is already full");
                        return;
                    }
                    break;
                case "medicine":
                    if (player.health < SETTINGS.max_health) {
                        playSoundEffect(soundEffects.medicine)
                        player.health = Math.min(player.health + SETTINGS.medicine_health, SETTINGS.max_health);
                    }else {
                        playSoundEffect(soundEffects.ERROR)
                        showSnackbar(snackbarType.ERROR, "Health is already full");
                        return;
                    }
                    break;
                case "food":
                    if (player.health < SETTINGS.max_health) {
                        playSoundEffect(soundEffects.food)
                        player.health = Math.min(player.health + SETTINGS.food_can_health, SETTINGS.max_health);
                    }else {
                        playSoundEffect(soundEffects.ERROR)
                        showSnackbar(snackbarType.ERROR, "Health is already full");
                        return;
                    }
                    break;
                case "water":
                    if (player.health < SETTINGS.max_health) {
                        playSoundEffect(soundEffects.water)
                        player.health = Math.min(player.health + SETTINGS.water_can_health, SETTINGS.max_health);
                    }
                    else {
                        playSoundEffect(soundEffects.ERROR)
                        showSnackbar(snackbarType.ERROR, "Health is already full");
                        return;
                    }
                    break;
                case "weapon":
                    playSoundEffect(soundEffects.ERROR)
                    showSnackbar(snackbarType.ERROR, "Weapon can't be used directly, they will automatically be used to defend against zombie attacks");
                    return;
                default:
                    playSoundEffect(soundEffects.ERROR)
                    showSnackbar(snackbarType.ERROR, "Unknown item");
                    return;
            }
            this.items[item] -= 1;
            showSnackbar(snackbarType.INFO,`Used ${item}. Remaining: ${this.items[item] ?? 0}`);
            updateUI(player)
        } else {
            playSoundEffect(soundEffects.ERROR);
            showSnackbar(snackbarType.ERROR, `${item} not available.`);
        }
    }

    generateRandomItems() {
        /** Generate random items at the start. */
        const randomInventoryItems = getRandomSubset(storableItems, 3);
        randomInventoryItems.forEach(item => {
            this.addItem(item, getRandomInt(1, 3));
        });
    }
}

// Helper functions
function getRandomSubset(array, size) {
    const shuffled = array.slice();
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
    }
    return shuffled.slice(0, size);
}

function getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

export function jsonifyInventory(inventory) {
    /** Convert inventory to a string for display. */
    return inventory

}